lib/npm.js

```diff
@@ -15,6 +16,42 @@ var Npmrc = require('./npmrc');
 
 var defaultRegistry = 'https://registry.npmjs.org';
 
+// Test whether the contents of buffer is gzipped
+function isGzip(buffer) {
+  if (!buffer || buffer.length < 3) {
+    return false;
+  }
+  return buffer[0] === 0x1f && buffer[1] === 0x8b && buffer[2] === 0x08;
+}
+
+// Return a promise that gunzips the buffer *if* it is gzipped
+// Otherwise return a promise with the buffer as is. 
+function gunzip(buffer) {
+  if (isGzip(buffer)) {
+    return new Promise(function (resolve, reject) {
+      zlib.gunzip(buffer, function (err, decoded) {
+        if (err) {
+          reject(err);
+        } else {
+          resolve(decoded);
+        }
+      });
+    });
+  } else {
+    return Promise.resolve(buffer);
+  }
+}
+
+// Create a readable stream from a buffer
+function bufferStream(buffer) {
+  return new stream.Readable({
+    read(size) {
+      this.push(buffer);
+      this.push(null);
+    }
```

Comment:
Well, I have not done any measurements, but the way I have done it in this PR the whole stream is read into memory as buffers and then concatenated. This must have some impact on performance, even though I don't quite know how big the impact is.

Using [node-buffer-peek-stream](https://github.com/seangarner/node-buffer-peek-stream) I guess there will be a small overhead in peeking at the first few bytes and "stuffing" them back into the stream, but we won't have the entire stream in memory, so I am pretty confident that this would be the more performant choice.